File: src/main/java/com/projeto/springapi/controller/ServidorEfetivoController.java
package com.projeto.springapi.controller;

import com.projeto.springapi.dto.ConsultaEnderecoFuncionalServidorEfetivoPorNomeDTO;
import com.projeto.springapi.dto.ConsultaServidorLotadoPorUnidadeDTO;
import com.projeto.springapi.dto.EnderecoDTO;
import com.projeto.springapi.dto.ServidorEfetivoDTO;
import com.projeto.springapi.model.ServidorEfetivo;
import com.projeto.springapi.service.ServidorEfetivoService;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/servidores-efetivos")
@Tag(name = "Servidores Efetivos", description = "Operações relacionadas a servidores efetivos")
public class ServidorEfetivoController {

    @Autowired
    private ServidorEfetivoService servidorEfetivoService;

    @Operation(summary = "Lista todos os servidores efetivos",
            description = "Retorna uma lista paginada de todos os servidores efetivos.",
            tags = {"Servidores Efetivos"})
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200",
                    description = "Lista de servidores efetivos retornada com sucesso",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = Page.class))),
            @ApiResponse(responseCode = "204", description = "Nenhum servidor efetivo encontrado",
                    content = @Content)})
    @GetMapping
    public ResponseEntity<Page<ServidorEfetivoDTO>> getAllServidoresEfetivos(@Parameter(
            description = "Objeto de paginação para controlar o número de elementos por página, a página atual e a ordenação",
            schema = @Schema(implementation = Pageable.class)) Pageable pageable) {
        Page<ServidorEfetivoDTO> servidores =
                servidorEfetivoService.getAllServidoresEfetivos(pageable);
        return servidores.isEmpty() ? ResponseEntity.noContent().build()
                : ResponseEntity.ok(servidores);
    }

    @Operation(summary = "Busca um servidor efetivo por ID",
            description = "Retorna os detalhes de um servidor efetivo específico com base no seu ID.",
            tags = {"Servidores Efetivos"})
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Servidor efetivo encontrado",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ServidorEfetivoDTO.class))),
            @ApiResponse(responseCode = "404", description = "Servidor efetivo não encontrado",
                    content = @Content)})
    @GetMapping("/{id}")
    public ResponseEntity<ServidorEfetivoDTO> getServidorEfetivoById(
            @Parameter(description = "ID do servidor efetivo a ser buscado",
                    required = true) @PathVariable Long id) {
        return ResponseEntity.ok(servidorEfetivoService.getServidorEfetivoById(id));
    }

    @Operation(summary = "Cria um novo servidor efetivo",
            description = "Cria um novo registro de servidor efetivo.",
            tags = {"Servidores Efetivos"})
    @ApiResponses(value = {
            @ApiResponse(responseCode = "201", description = "Servidor efetivo criado com sucesso",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ServidorEfetivoDTO.class))),
            @ApiResponse(responseCode = "400", description = "Dados de entrada inválidos",
                    content = @Content)})
    @PostMapping
    public ResponseEntity<ServidorEfetivoDTO> createServidorEfetivo(@Parameter(
            description = "Dados do servidor efetivo a serem criados", required = true,
            schema = @Schema(
                    implementation = ServidorEfetivoDTO.class)) @RequestBody ServidorEfetivoDTO servidorEfetivoDTO) {
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(servidorEfetivoService.createServidorEfetivo(servidorEfetivoDTO));
    }

    @Operation(summary = "Atualiza um servidor efetivo existente",
            description = "Atualiza os dados de um servidor efetivo específico com base no seu ID.",
            tags = {"Servidores Efetivos"})
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200",
                    description = "Servidor efetivo atualizado com sucesso",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ServidorEfetivoDTO.class))),
            @ApiResponse(responseCode = "400", description = "Dados de entrada inválidos",
                    content = @Content),
            @ApiResponse(responseCode = "404", description = "Servidor efetivo não encontrado",
                    content = @Content)})
    @PutMapping("/{id}")
    public ResponseEntity<ServidorEfetivoDTO> updateServidorEfetivo(
            @Parameter(description = "ID do servidor efetivo a ser atualizado",
                    required = true) @PathVariable Long id,
            @Parameter(description = "Novos dados do servidor efetivo", required = true,
                    schema = @Schema(
                            implementation = ServidorEfetivoDTO.class)) @RequestBody ServidorEfetivoDTO servidorEfetivoDTO) {
        return ResponseEntity
                .ok(servidorEfetivoService.updateServidorEfetivo(id, servidorEfetivoDTO));
    }

    @Operation(summary = "Exclui um servidor efetivo por ID",
            description = "Remove um servidor efetivo específico com base no seu ID.",
            tags = {"Servidores Efetivos"})
    @ApiResponses(value = {
            @ApiResponse(responseCode = "204",
                    description = "Servidor efetivo excluído com sucesso", content = @Content),
            @ApiResponse(responseCode = "404", description = "Servidor efetivo não encontrado",
                    content = @Content)})
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteServidorEfetivo(
            @Parameter(description = "ID do servidor efetivo a ser excluído",
                    required = true) @PathVariable Long id) {
        servidorEfetivoService.deleteServidorEfetivo(id);
        return ResponseEntity.noContent().build();
    }

    @Operation(summary = "Adiciona um endereço a um servidor efetivo existente",
            description = "Adiciona um novo endereço à servidor efetivo especificada.",
            tags = {"Servidores Efetivos", "Endereços"})
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Endereço adicionado com sucesso",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ServidorEfetivo.class))),
            @ApiResponse(responseCode = "404", description = "Servidor Efetivo não encontrado",
                    content = @Content),
            @ApiResponse(responseCode = "400", description = "Requisição inválida",
                    content = @Content)})
    @PostMapping("/{id}/enderecos")
    public ResponseEntity<ServidorEfetivoDTO> addEnderecoToServidorEfetivo(
            @Parameter(description = "ID do servidor efetivo para adicionar o endereço",
                    required = true) @PathVariable Long id,
            @Parameter(description = "Dados do endereço a serem adicionados", required = true,
                    schema = @Schema(
                            implementation = EnderecoDTO.class)) @RequestBody EnderecoDTO enderecoDTO) {
        return ResponseEntity
                .ok(servidorEfetivoService.addEnderecoToServidorEfetivo(id, enderecoDTO));
    }

    @Operation(summary = "Lista os servidores efetivos por ID da unidade com paginação",
            description = "Retorna uma página de servidores efetivos pertencentes a uma unidade específica.",
            tags = {"Busca"})
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200",
                    description = "Página de servidores efetivos da unidade retornada com sucesso",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(
                                    implementation = ConsultaServidorLotadoPorUnidadeDTO.class))),
            @ApiResponse(responseCode = "204",
                    description = "Nenhum servidor efetivo encontrado para esta unidade nesta página",
                    content = @Content),
            @ApiResponse(responseCode = "404", description = "Unidade não encontrada",
                    content = @Content)})
    @GetMapping("/unidade/{unidadeId}")
    public ResponseEntity<Page<ConsultaServidorLotadoPorUnidadeDTO>> getServidoresEfetivosByUnidade(
            @Parameter(description = "ID da unidade para buscar os servidores efetivos",
                    required = true) @PathVariable Long unidadeId,
            Pageable pageable) {
        Page<ConsultaServidorLotadoPorUnidadeDTO> consulta =
                servidorEfetivoService.findServidoresEfetivosLotadosPorUnidade(unidadeId, pageable);
        return consulta.isEmpty() ? ResponseEntity.noContent().build()
                : ResponseEntity.ok(consulta);
    }

    @Operation(
            summary = "Consulta paginada de endereço funcional de servidores efetivos por nome (parcial ou completo)",
            description = "Retorna uma página de endereços de unidades de servidores efetivos cujo nome corresponde ao parâmetro de busca.",
            tags = {"Busca"})
    @ApiResponses(value = {@ApiResponse(responseCode = "200",
            description = "Página de endereços funcionais de servidores efetivos encontrados por nome retornada com sucesso",
            content = @Content(mediaType = "application/json", schema = @Schema(
                    implementation = ConsultaEnderecoFuncionalServidorEfetivoPorNomeDTO.class))),
            @ApiResponse(responseCode = "204",
                    description = "Nenhum servidor efetivo encontrado com este nome nesta página",
                    content = @Content)})
    @GetMapping("/endereco-funcional")
    public ResponseEntity<Page<ConsultaEnderecoFuncionalServidorEfetivoPorNomeDTO>> getEnderecoFuncionalByServidorNomeContaining(
            @Parameter(description = "Nome (ou parte do nome) do servidor efetivo a ser buscado",
                    required = true) @RequestParam("nome") String nome,
            Pageable pageable) {
        Page<ConsultaEnderecoFuncionalServidorEfetivoPorNomeDTO> servidores =
                servidorEfetivoService.getEnderecoFuncionalByServidorNomeContaining(nome, pageable);
        return servidores.isEmpty() ? ResponseEntity.noContent().build()
                : ResponseEntity.ok(servidores);
    }
}

File: app/Http/Controllers/ServidorEfetivoController.php
<?php

namespace App\Http\Controllers;

use App\Models\Endereco;
use App\Models\Pessoa;
use App\Models\ServidorEfetivo;
use App\Http\Resources\ServidorEfetivoResource;
use App\Http\Resources\ConsultaServidorLotadoPorUnidadeResource;
use App\Http\Resources\ConsultaEnderecoFuncionalServidorEfetivoPorNomeResource;
use App\Http\Requests\StoreServidorEfetivoRequest;
use App\Http\Requests\UpdateServidorEfetivoRequest;
use App\Http\Requests\AddEnderecoToServidorEfetivoRequest;
use DB;
use Illuminate\Http\Request;
use Illuminate\Http\Response;

class ServidorEfetivoController extends Controller
{
    /**
     * @OA\Get(
     *     path="/api/servidores-efetivos",
     *     tags={"Servidores Efetivos"},
     *     summary="Lista todos os servidores efetivos",
     *     description="Retorna uma lista paginada de todos os servidores efetivos.",
     *     security={{"sanctum": {}}},
     *     @OA\Response(
     *         response=200,
     *         description="Lista de servidores efetivos retornada com sucesso",
     *         @OA\JsonContent(
     *             type="array",
     *             @OA\Items(ref="#/components/schemas/ServidorEfetivoResource")
     *         )
     *     ),
     *     @OA\Response(
     *         response=204,
     *         description="Nenhum servidor efetivo encontrado"
     *     )
     * )
     */
    public function index(Request $request)
    {
        $perPage = $request->input('per_page', 10);
        $servidores = ServidorEfetivo::with(['pessoa', 'pessoa.enderecos'])->paginate($perPage);
        
        if ($servidores->isEmpty()) {
            return response()->noContent();
        }
        
        return ServidorEfetivoResource::collection($servidores);
    }

    /**
     * @OA\Get(
     *     path="/api/servidores-efetivos/{id}",
     *     tags={"Servidores Efetivos"},
     *     summary="Busca um servidor efetivo por ID",
     *     description="Retorna os detalhes de um servidor efetivo específico com base no seu ID.",
     *     security={{"sanctum": {}}},
     *     @OA\Parameter(
     *         name="id",
     *         in="path",
     *         description="ID do servidor efetivo a ser buscado",
     *         required=true,
     *         @OA\Schema(type="integer")
     *     ),
     *     @OA\Response(
     *         response=200,
     *         description="Servidor efetivo encontrado",
     *         @OA\JsonContent(ref="#/components/schemas/ServidorEfetivoResource")
     *     ),
     *     @OA\Response(
     *         response=404,
     *         description="Servidor efetivo não encontrado"
     *     )
     * )
     */
    public function show($id)
    {
        $servidor = ServidorEfetivo::with(['pessoa', 'pessoa.enderecos'])->findOrFail($id);
        return new ServidorEfetivoResource($servidor);
    }

    /**
     * @OA\Post(
     *     path="/api/servidores-efetivos",
     *     tags={"Servidores Efetivos"},
     *     summary="Cria um novo servidor efetivo",
     *     description="Cria um novo registro de servidor efetivo.",
     *     security={{"sanctum": {}}},
     *     @OA\RequestBody(
     *         required=true,
     *         @OA\JsonContent(ref="#/components/schemas/StoreServidorEfetivoRequest")
     *     ),
     *     @OA\Response(
     *         response=201,
     *         description="Servidor efetivo criado com sucesso",
     *         @OA\JsonContent(ref="#/components/schemas/ServidorEfetivoResource")
     *     ),
     *     @OA\Response(
     *         response=400,
     *         description="Dados de entrada inválidos"
     *     )
     * )
     */
    public function store(StoreServidorEfetivoRequest $request)
    {
        $validated = $request->validated();

        $response = DB::transaction(function () use ($validated) {
            $pessoa = Pessoa::create([
                'pes_nome'             => $validated['pes_nome'],
                'pes_data_nascimento'  => $validated['pes_data_nascimento'],
                'pes_sexo'             => $validated['pes_sexo'],
                'pes_mae'              => $validated['pes_mae'],
                'pes_pai'              => $validated['pes_pai'],
            ]);
    
            $servidor = ServidorEfetivo::create([
                'pes_id'       => $pessoa->pes_id,
                'se_matricula' => $validated['se_matricula'],
            ]);
    
            if (!empty($validated['endereco_ids'])) {
                $pessoa->enderecos()->sync($validated['endereco_ids']);
            }
    
            return response()->json($servidor->load('pessoa.enderecos'), 201);
        });

        return $response;
    }

    /**
     * @OA\Put(
     *     path="/api/servidores-efetivos/{id}",
     *     tags={"Servidores Efetivos"},
     *     summary="Atualiza um servidor efetivo existente",
     *     description="Atualiza os dados de um servidor efetivo específico com base no seu ID.",
     *     security={{"sanctum": {}}},
     *     @OA\Parameter(
     *         name="id",
     *         in="path",
     *         description="ID do servidor efetivo a ser atualizado",
     *         required=true,
     *         @OA\Schema(type="integer")
     *     ),
     *     @OA\RequestBody(
     *         required=true,
     *         @OA\JsonContent(ref="#/components/schemas/UpdateServidorEfetivoRequest")
     *     ),
     *     @OA\Response(
     *         response=200,
     *         description="Servidor efetivo atualizado com sucesso",
     *         @OA\JsonContent(ref="#/components/schemas/ServidorEfetivoResource")
     *     ),
     *     @OA\Response(
     *         response=400,
     *         description="Dados de entrada inválidos"
     *     ),
     *     @OA\Response(
     *         response=404,
     *         description="Servidor efetivo não encontrado"
     *     )
     * )
     */
    public function update(UpdateServidorEfetivoRequest $request, $id)
    {
        $servidor = ServidorEfetivo::findOrFail($id);
        $servidor->update($request->validated());
        
        if ($request->has('endereco_ids')) {
            $servidor->pessoa->enderecos()->sync($request->endereco_ids);
        }
        
        return new ServidorEfetivoResource($servidor);
    }

    /**
     * @OA\Delete(
     *     path="/api/servidores-efetivos/{id}",
     *     tags={"Servidores Efetivos"},
     *     summary="Exclui um servidor efetivo por ID",
     *     description="Remove um servidor efetivo específico com base no seu ID.",
     *     security={{"sanctum": {}}},
     *     @OA\Parameter(
     *         name="id",
     *         in="path",
     *         description="ID do servidor efetivo a ser excluído",
     *         required=true,
     *         @OA\Schema(type="integer")
     *     ),
     *     @OA\Response(
     *         response=204,
     *         description="Servidor efetivo excluído com sucesso"
     *     ),
     *     @OA\Response(
     *         response=404,
     *         description="Servidor efetivo não encontrado"
     *     )
     * )
     */
    public function destroy($id)
    {
        $servidor = ServidorEfetivo::findOrFail($id);
        $servidor->delete();
        
        return response()->noContent();
    }

    /**
     * @OA\Post(
     *     path="/api/servidores-efetivos/{id}/enderecos",
     *     tags={"Servidores Efetivos", "Endereços"},
     *     summary="Adiciona um endereço a um servidor efetivo existente",
     *     description="Adiciona um novo endereço à servidor efetivo especificada.",
     *     security={{"sanctum": {}}},
     *     @OA\Parameter(
     *         name="id",
     *         in="path",
     *         description="ID do servidor efetivo para adicionar o endereço",
     *         required=true,
     *         @OA\Schema(type="integer")
     *     ),
     *     @OA\RequestBody(
     *         required=true,
     *         @OA\JsonContent(ref="#/components/schemas/AddEnderecoToServidorEfetivoRequest")
     *     ),
     *     @OA\Response(
     *         response=200,
     *         description="Endereço adicionado com sucesso",
     *         @OA\JsonContent(ref="#/components/schemas/ServidorEfetivoResource")
     *     ),
     *     @OA\Response(
     *         response=404,
     *         description="Servidor Efetivo não encontrado"
     *     ),
     *     @OA\Response(
     *         response=400,
     *         description="Requisição inválida"
     *     )
     * )
     */
    public function addEndereco(AddEnderecoToServidorEfetivoRequest $request, $id)
    {
        $servidor = ServidorEfetivo::findOrFail($id);
        $endereco = Endereco::create($request->validated());

        $servidor->pessoa->enderecos()->attach($endereco->end_id);

        return new ServidorEfetivoResource($servidor->load('pessoa.enderecos'));
    }

    /**
     * @OA\Get(
     *     path="/api/servidores-efetivos/unidade/{unidadeId}",
     *     tags={"Busca"},
     *     summary="Lista os servidores efetivos por ID da unidade",
     *     description="Retorna uma lista de servidores efetivos pertencentes a uma unidade específica.",
     *     security={{"sanctum": {}}},
     *     @OA\Parameter(
     *         name="unidadeId",
     *         in="path",
     *         description="ID da unidade para buscar os servidores efetivos",
     *         required=true,
     *         @OA\Schema(type="integer")
     *     ),
     *     @OA\Response(
     *         response=200,
     *         description="Lista de servidores efetivos da unidade retornada com sucesso",
     *         @OA\JsonContent(
     *             type="array",
     *             @OA\Items(ref="#/components/schemas/ConsultaServidorLotadoPorUnidadeResource")
     *         )
     *     ),
     *     @OA\Response(
     *         response=204,
     *         description="Nenhum servidor efetivo encontrado para esta unidade"
     *     ),
     *     @OA\Response(
     *         response=404,
     *         description="Unidade não encontrada"
     *     )
     * )
     */
    public function getByUnidade($unidadeId)
    {
        $servidores = ServidorEfetivo::whereHas('lotacoes', function($query) use ($unidadeId) {
            $query->where('unid_id', $unidadeId);
        })->with(['pessoa', 'lotacoes.unidade'])->paginate(10);
        
        if ($servidores->isEmpty()) {
            return response()->noContent();
        }
        
        return ConsultaServidorLotadoPorUnidadeResource::collection($servidores);
    }

    /**
     * @OA\Get(
     *     path="/api/servidores-efetivos/endereco-funcional",
     *     tags={"Busca"},
     *     summary="Consulta endereço funcional de servidores efetivos por nome (parcial ou completo)",
     *     description="Retorna uma lista de endereços de unidades de servidores efetivos cujo nome corresponde ao parâmetro de busca.",
     *     security={{"sanctum": {}}},
     *     @OA\Parameter(
     *         name="nome",
     *         in="query",
     *         description="Nome (ou parte do nome) do servidor efetivo a ser buscado",
     *         required=true,
     *         @OA\Schema(type="string")
     *     ),
     *     @OA\Response(
     *         response=200,
     *         description="Lista de endereços funcionais de servidores efetivos encontrados por nome retornada com sucesso",
     *         @OA\JsonContent(
     *             type="array",
     *             @OA\Items(ref="#/components/schemas/ConsultaEnderecoFuncionalServidorEfetivoPorNomeResource")
     *         )
     *     ),
     *     @OA\Response(
     *         response=204,
     *         description="Nenhum servidor efetivo encontrado com este nome"
     *     )
     * )
     */
    public function getEnderecoFuncionalByNome(Request $request)
    {
        $nome = $request->query('nome');
        $servidores = ServidorEfetivo::whereHas('pessoa', function($query) use ($nome) {
            $query->where('pes_nome', 'like', "%{$nome}%");
        })->with(['pessoa', 'lotacoes.unidade.enderecos'])->paginate(10);

        if ($servidores->isEmpty()) {
            return response()->noContent();
        }

        return ConsultaEnderecoFuncionalServidorEfetivoPorNomeResource::collection($servidores);
    }
}
